### ISSUE TITLE
Feature request: GitLab forkMode

### ISSUE THREAD

**Issue by Phyks** [renovatebot/renovate#840]
Hi,

I'd like to use `renovate` on a project hosted on Gitlab. I went through the doc and managed to set it up on my project, but it seems it requires a `developer` permission on the project, to be able to create branches.

Would it be possible to make it create MR from its own fork, to avoid granting too much permissions to the bot account?

Thanks!

**Comment by rarkins** [comment]
This is not currently possible and is not something I have planned to do. However if someone scoped how much work it would be then I'd consider merging a PR if it wouldn't impact the existing architecture too much.

Aren't you running the Renovate bot account though? Ie you are only really assigning privileges to yourself anyway?

**Comment by Phyks** [comment]
Ok, thanks for the very quick answer on this!

I am running an account for the bot, so indeed I control the bot behavior. However, we never push directly to the repo, but rather do MR with a merge without merge commit. So, if I could give the bot access to its own fork, and no access to my repo besides creating merge requests, it would have enforced this behavior.

I totally agree there should not be any issue with it, I just wanted to be as safe as possible with any bug :)

**Comment by rarkins** [comment]
Thanks for the explanation. I'm most familiar with the GitHub approach where you gave good controls for limiting if/when accounts can merge to base branch. Eg only after PR, or only if branch passes all tests , etc. so this essentially achieves what you describe

**Comment by rarkins** [comment]
These GitLab settings could help alleviate some of your concerns about the bot going buggy at some point: https://docs.gitlab.com/ee/user/project/protected_branches.html

**Comment by Phyks** [comment]
We already set something like this. But we can only target "permissions levels", not individual developers. Moving to "masters only" would block other contributors. But sure, this is more a problem with Gitlab limitations than a real issue with renovate.

**Comment by rarkins** [comment]
Closing this with label `wontfix` but of course it can be reopened if there is demand for it in future.

**Comment by Phyks** [comment]
Another interesting use case I just noticed is that this way you don't have to worry at all about `renovate/*` branches eventually left behind after a merge and polluting your branches namespace in your own project.

That being said, I don't think I would have time to submit a MR to handle this soon, and this is definitely not something which would have high priority, so I am totally ok for closing this as `wontfix`. Thanks for your work on `renovate`!

**Comment by rarkins** [comment]
@Phyks FYI, Renovate already cleans up `renovate/*` branches in GitHub, so the functionality can be ported to GitLab (I forget what held it back initially but I think now it's solved).

**Comment by rarkins** [comment]
I've reopened this as I've been asked about this capability again, for organisations that have policies against granting write access to repositories. I will plan this from an App perspective but then implement in a generic way that satisfies self-hosted too.

For a public project, it should not require any permissions in order to create PRs from the bot's own fork.

Keeping the fork updated and non-conflicted can be a challenge though. We can't delete the fork between "runs" because there may still be PRs outstanding that would then lose their reference. But at the same time the master in the source repo may have dramatic changes between runs too.

The fork's *master* branch should never be updated though - only `renovate/*` branches. And unless the source repo does force pushes to `master`, `git merge` should always be fast-forward?

Possible it works with just these changes?

#### Fork and/or refresh fork `master` branch

If some "useForkMode" option is set to true for the repo, then when renovate starts on that repo it will:
  - Check if fork exists already. If not then forks it, clones it, and adds upstream reference to source. Need to consider namespace conflicts though (e.g. two separate accounts with "foo" repository)
  - If exists, then does `git merge` from upstream and pushes to github/gitlab

#### Run like normal but using the forked repo instead

e.g. use content from forked master, keep forked branches up-to-date, etc

#### Don't use own master as target of PR, instead use upstream repo as target

PR get/set operations should target source repo, not the fork.

GitHub docs: 
- https://developer.github.com/v3/repos/forks/#create-a-fork
- https://help.github.com/articles/syncing-a-fork/

**Comment by rarkins** [comment]
When forking through GitHub's web UI, it automatically avoids namespace collisions by adding a number suffix, e.g. it would all it `foo-1` if `foo` already exists.

**Comment by rarkins** [comment]
This would need to be an opt-in too, as we wouldn't want people *accidentally* submitting PRs to public projects when they just intended to test or had typo'd;

**Comment by Phyks** [comment]
Nice to see it reopened! Thanks for any work on this! :)

**Comment by rarkins** [comment]
Requirements:

- [ ] New config option `forkRepo` defaulting to `false` (in `renovate.json`)
- [ ] Renovate checks if it already has a fork of the repo. If not then it forks via API and then sleeps/tries/sleeps/tries until it shows up (with some limit just in case)
- [ ] Once fork exists, then it needs to be updated so that `master` sha matches with source (ideally use `baseBranch` not hardcoded `master`
- [ ] Run as normal using repository fork, but the PR target needs to be source:master and not fork:master

**Comment by rarkins** [comment]
GitLab support for this is unfortunately not possible at this time so is not included in PR #1287. Although it's possible to create forks using the GitLab API, I cannot find any feasible way to keep the fork's base branch up to date with the source, like is done with GitHub using the raw git api.

References:
- https://github.com/samrocketman/gitlab-mirrors
- https://docs.gitlab.com/ee/workflow/repository_mirroring.html
- https://about.gitlab.com/2016/12/01/how-to-keep-your-fork-up-to-date-with-its-origin/

**Comment by rarkins** [comment]
Sorry, but it has not been implemented. GitLab so far provide no way to programmatically "sync" a fork's master with its parent. If there ever is, then we can reopen.

**Comment by kevincox** [comment]
I'm curious why you need to sync master? Shouldn't you just create a new branch based off of the upstream master? IIUC the forked repo doesn't even really need a master branch.

**Comment by rarkins** [comment]
I have no objection if want to try implementing it that way. Originally we weren't using raw git so didn't have that level of control.

**Comment by viceice** [comment]
still needs discussion how to since the fork base branches with the upstream ones. should ideally be done without a local copy

**Comment by rarkins** [comment]
Blocked by #13586

**Comment by RahulGautamSingh** [comment]
> Blocked by [#13586](https://github.com/renovatebot/renovate/issues/13586)

The blocker has been closed and I think the issue also resolved now since we control at git level and platform might not matter as much

**Comment by rarkins** [comment]
I think there's more work needed

**Comment by RahulGautamSingh** [comment]
Yes. I missed that we need to make some changes to respective the platform modules as well.

**Comment by ltickett** [comment]
I have a merge request open providing a mechanism to grant renovate developer access to a fork only: https://github.com/renovatebot/renovate/pull/39246

At the moment, you would have to create the fork manually (or using terraform etc). 
Hopefully we can implement automatic forking in another PR.

Please feel free to take a look at the solution- thanks

**Comment by RahulGautamSingh** [comment]
> Hopefully we can implement automatic forking in another PR.

Does this mean it can be done, or to-be-done when Gitlab provides endpoint to do that?

**Comment by ltickett** [comment]
Yes @RahulGautamSingh - we can definitely tackle the automatic forking later too (GitLab does support that) 👍

**Comment by RahulGautamSingh** [comment]
Then, lets do that. As we can handle syncing with upstream branch using our git layer.

**Comment by ltickett** [comment]
Do you mean handling automatic forking first as opposed to https://github.com/renovatebot/renovate/pull/39246

**Comment by RahulGautamSingh** [comment]
Yes, if we can create a fork and on subsequent runs use the existing fork(created by bot in previous run) that makes more sense to have than asking user to create a fork

**Comment by viceice** [comment]
we should refactor renovate first to extract the general fork handling to be first class citizen equal for all platforms, the each platform can implement the specific pieces needed.

that way we don't need to duplicate and maintain too much complicated code for each platform.

see this discussion for general agreement 
- https://github.com/renovatebot/renovate/discussions/39263
